runtime/doc/lua-guide.txt

```diff
@@ -0,0 +1,756 @@
+functions accessible through `vim.*` not mentioned already.
+
+This distinction is important, as API functions inherit behavior from their
+original layer: For example, Neovim API functions always need all arguments to
+be specified even if Lua itself allows omitting arguments (which are then
+passed as `nil`); and Vim API functions can use 0-based indexing even if Lua
+arrays are 1-indexed by default.
+
+Through this, any possible interaction can be done through Lua without writing
+a complete new API from scratch. For this reason, functions are usually not
+duplicated between layers unless there is a significant benefit in
+functionality or performance (e.g., you can map Lua functions directly through
+|nvim_create_autocmd| but not through |:autocmd|). In case there are multiple
+ways of achieving the same thing, this guide will only cover what is most
+convenient to from Lua.
+
+==============================================================================
+Using Lua                                                  *lua-guide-using-Lua*
+
+To run Lua code from the Neovim command line, use the |:lua| command:
+>vim
+    :lua print("Hello!")
+<
+Note: each `:lua` command has its own scope and variables declared with the
+local keyword are not accessible outside of the command. This won't work:
+>vim
+    :lua local foo = 1
+    :lua print(foo)
+    " prints "nil" instead of "1"
+<
+You can also use `:lua =`, which is same as `:lua print(vim.inspect(expr))`, to
+conveniently check the value of a variable or a table:
+>lua
+    :lua =package
+<
+To run a Lua script in an external file, you can use the |:source| command
+exactly like for a Vimscript:
+>vim
+    :source ~/programs/baz/myluafile.lua
+<
+Finally, you can include Lua code in a vimscript file by putting it inside a
+|lua-heredoc| block:
+>vim
+    lua << EOF
+        local tbl = {1, 2, 3}
+        for k, v in ipairs(tbl) do
+            print(v)
+        end
+    EOF
+<
+------------------------------------------------------------------------------
+Using Lua files on startup                                    *lua-guide-config*
+
+Neovim supports using `init.vim` or `init.lua` as the configuration file, but
+not both at the same time. This should be placed in your |config| directory,
+which is typically `~/.config/nvim` for Linux, BSD, or macOS, and
+`~/AppData/Local/nvim/` for Windows. Note that you can use Lua in `init.vim`
+and Vimscript in `init.lua`, which will be covered below.
+
+If you'd like to run any other Lua script on |startup| automatically, then you
+can simply put it in `plugin/` in your |'runtimepath'|.
+
+------------------------------------------------------------------------------
+Lua modules                                                  *lua-guide-modules*
+
+If you want to load Lua files on demand, you can place them in the `lua/`
+directory in your |'runtimepath'| and load them with `require`. (This is the
+Lua equivalent of Vimscript's |autoload|.)
```

Comment:
```suggestion
directory in your |'runtimepath'| and load them with `require()`. This is similar 
to the Vim script |autoload| mechanism.
```

Unless it's exactly the same, which I don't think it is, I don't think we should say 'equivalent' here,  aren't there lots of situations and small nuances between the two loading mechanisms?

Reply:
The main point (on demand vs. on startup) is indeed exactly the same.